File: app/src/test/java/com/omar/retromp3recorder/app/player/StateLoggingAudioPlayerTest.java
package com.omar.retromp3recorder.app.player;

import com.omar.retromp3recorder.app.di.DaggerTestAppComponent;
import com.omar.retromp3recorder.app.main.MainView;
import com.omar.retromp3recorder.app.repo.PlayerIdRepo;
import com.omar.retromp3recorder.app.repo.StateRepo;

import org.junit.Before;
import org.junit.Test;

import javax.inject.Inject;
import javax.inject.Named;

import io.reactivex.Scheduler;
import io.reactivex.subjects.Subject;

import static com.omar.retromp3recorder.app.di.AppComponent.DECORATOR_ALPHA;
import static com.omar.retromp3recorder.app.di.MockModule.PLAYER_SUBJECT;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.verify;

public class StateLoggingAudioPlayerTest {


    @Inject
    StateRepo stateRepo;
    @Inject
    PlayerIdRepo playerIdRepo;
    @Inject
    Scheduler scheduler;
    @Inject
    @Named(DECORATOR_ALPHA)
    AudioPlayer baseAudioPlayer;
    @Inject
    @Named(PLAYER_SUBJECT)
    Subject<AudioPlayer.Event> audioEvents;
    private AudioPlayer spy;
    private StateLoggingAudioPlayer stateLoggingAudioPlayer;

    @Before
    public void setUp() {
        DaggerTestAppComponent.create().inject(this);
        spy = spy(baseAudioPlayer);
        stateLoggingAudioPlayer = new StateLoggingAudioPlayer(stateRepo, spy, playerIdRepo, scheduler);
    }

    @Test
    public void test_DidDecorateStartPlay() {
        stateLoggingAudioPlayer.playerStart("test");

        verify(spy).playerStart("test");
    }

    @Test
    public void test_DidDecorateStopPlay() {
        stateLoggingAudioPlayer.playerStop();

        verify(spy).playerStop();
    }

    @Test
    public void test_OnStartPlay_PostState() {
        stateLoggingAudioPlayer.playerStart("test");

        //Then
        stateRepo.observe()
                .test()
                .assertValue(state -> state == MainView.State.Playing);
    }

    @Test
    public void test_OnStopPlayPostState(){
        stateLoggingAudioPlayer.playerStop();

        //Then
        stateRepo.observe()
                .test()
                .assertValue(state -> state == MainView.State.Idle);
    }

    @Test
    public void test_PlayerId_PostToRepo() {
        audioEvents.onNext(new AudioPlayer.SendPlayerId(10));

        playerIdRepo.observe().test()
                .assertValue(integerOneShot ->
                        integerOneShot.getValueOnce() == 10
                );
    }

    @Test
    public void test_OnPlayerError_PostIdleState() {
        audioEvents.onNext(new AudioPlayer.Error("test"));

        stateRepo.observe().test()
                .assertValue(state ->
                        state == MainView.State.Idle
                );
    }

    @Test
    public void test_DidDecorateIsPlaying(){
        assert stateLoggingAudioPlayer.isPlaying();
    }
}

File: app/src/test/java/com/omar/retromp3recorder/app/playback/player/StateLoggingAudioPlayerTest.kt
package com.omar.retromp3recorder.app.playback.player

import com.github.alkurop.ghostinshell.Shell
import com.github.alkurop.stringerbell.Stringer.Companion.ofString
import com.omar.retromp3recorder.app.common.repo.StateRepo
import com.omar.retromp3recorder.app.di.AppComponent
import com.omar.retromp3recorder.app.di.DaggerTestAppComponent
import com.omar.retromp3recorder.app.di.MockModule
import com.omar.retromp3recorder.app.main.MainView
import com.omar.retromp3recorder.app.playback.player.AudioPlayer.Event.SendPlayerId
import com.omar.retromp3recorder.app.playback.repo.PlayerIdRepo
import io.reactivex.Scheduler
import io.reactivex.subjects.Subject
import org.junit.Before
import org.junit.Test
import org.mockito.Mockito
import javax.inject.Inject
import javax.inject.Named

class StateLoggingAudioPlayerTest {
    @Inject
    lateinit var stateRepo: StateRepo

    @Inject
    lateinit var playerIdRepo: PlayerIdRepo

    @Inject
    lateinit var scheduler: Scheduler

    @Inject
    @Named(AppComponent.DECORATOR_A)
    lateinit var baseAudioPlayer: AudioPlayer

    @Inject
    @Named(MockModule.PLAYER_SUBJECT)
    lateinit var audioEvents: Subject<AudioPlayer.Event>
    private lateinit var spy: AudioPlayer
    private lateinit var stateLoggingAudioPlayer: StateLoggingAudioPlayer

    @Before
    fun setUp() {
        DaggerTestAppComponent.create().inject(this)
        spy = Mockito.spy(baseAudioPlayer)
        stateLoggingAudioPlayer = StateLoggingAudioPlayer(
            stateRepo, spy,
            playerIdRepo, scheduler
        )
    }

    @Test
    fun test_DidDecorateStartPlay() {
        stateLoggingAudioPlayer.playerStart("test")

        //Then
        Mockito.verify(spy).playerStart("test")
    }

    @Test
    fun test_DidDecorateStopPlay() {
        stateLoggingAudioPlayer.playerStop()

        //Then
        Mockito.verify(spy).playerStop()
    }

    @Test
    fun test_OnStartPlay_PostState() {
        stateLoggingAudioPlayer.playerStart("test")

        //Then
        stateRepo.observe()
            .test()
            .assertValue { state: MainView.State -> state === MainView.State.Playing }
    }

    @Test
    fun test_OnStopPlayPostState() {
        stateLoggingAudioPlayer.playerStop()

        //Then
        stateRepo.observe()
            .test()
            .assertValue { state: MainView.State -> state === MainView.State.Idle }
    }

    @Test
    fun test_PlayerId_PostToRepo() {
        audioEvents.onNext(SendPlayerId(10))

        //Then
        playerIdRepo.observe().test()
            .assertValue { integerOneShot: Shell<Int> -> integerOneShot.ghost == 10 }
    }

    @Test
    fun test_OnPlayerError_PostIdleState() {
        audioEvents.onNext(AudioPlayer.Event.Error(ofString("test")))

        //Then
        stateRepo.observe().test()
            .assertValue { state: MainView.State -> state === MainView.State.Idle }
    }

    @Test
    fun test_DidDecorateIsPlaying() {
        assert(stateLoggingAudioPlayer.isPlaying)
    }
}